lib/collection/src/operations/snapshot_s3_ops.rs

```diff
@@ -0,0 +1,238 @@
+use std::fs::File;
+use std::io::Write;
+use std::path::Path;
+
+use aws_sdk_s3::operation::complete_multipart_upload::CompleteMultipartUploadOutput;
+use aws_sdk_s3::operation::create_multipart_upload::CreateMultipartUploadOutput;
+use aws_sdk_s3::types::{CompletedMultipartUpload, CompletedPart};
+use aws_smithy_types::byte_stream::{ByteStream, Length};
+use aws_smithy_types_convert::date_time::DateTimeExt;
+
+use super::snapshot_ops::SnapshotDescription;
+use super::types::{CollectionError, CollectionResult};
+
+pub fn get_key(path: &Path) -> CollectionResult<String> {
+    // Get file name by trimming the path.
+    // if the path is ./path/to/file.txt, the key should be path/to/file.txt
+    let key = path.to_str().ok_or_else(|| {
+        CollectionError::s3_error(format!(
+            "Failed to get key for snapshot: {}",
+            path.display()
+        ))
+    });
+    key.map(|k| k.trim_start_matches("./").to_string())
+}
+
+pub async fn multi_part_upload(
+    client: &aws_sdk_s3::Client,
+    bucket_name: &str,
+    key: &str,
+    source_path: &str,
+) -> CollectionResult<CompleteMultipartUploadOutput> {
+    /// Performs a multipart upload to AWS S3.
+    ///
+    /// Steps for multipart upload:
+    /// 1. Initiate the upload by requesting an `upload_id` from AWS S3. This ID will uniquely identify the multipart upload session.
+    /// 2. Split the file into multiple chunks. Each chunk can be uploaded in parallel, which is useful for large files.
+    /// 3. Upload each chunk to AWS S3 using the `upload_id`. Each chunk is referred to as a part, and each part is uploaded separately.
+    /// 4. Once all parts have been uploaded, send a `complete_multipart_upload` request to AWS S3. This request will assemble all the uploaded parts into a single file based on the `upload_id`.
+    ///
+    /// This function handles each of these steps and includes error handling for each part of the process.
+    const CHUNK_SIZE: u64 = 1024 * 1024 * 50; // 50MB
+    const MAX_CHUNKS: u64 = 10000;
+
```

Comment:
I wasn't sure if this should be configured from config.yml. So i'm leaving these here for now.